### PR TITLE
feat: Support Transfer render

### PR DESCRIPTION
**Pull request by zombieJ**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### 👻 What's the background?

ref: #14705

### 💡 Solution

Official support of customize list render

### 📝 Changelog

Transfer support `children` prop to customize render list.

### ☑️ Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----
[View rendered components/transfer/demo/customize.md](https://github.com/ant-design/ant-design/blob/transfer-render/components/transfer/demo/customize.md)

### CODE REVIEW

**Comment by afc163** [comment]
This pull request **introduces 4 alerts** when merging 00a507bc2d62042f1e17ea21a1cc83783cca762b into c4283ebe72fd79b17f9c5ff596c0a83d177cd4c5 - [view on LGTM.com](https://lgtm.com/projects/g/ant-design/ant-design/rev/pr-122b08c600036640743979439cab60b5a753d929)

**new alerts:**

* 4 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by afc163** [comment]
This pull request **introduces 4 alerts** when merging 52c6e97fa8eb654afdb53a36a8b1ac559ba5e3e2 into c4283ebe72fd79b17f9c5ff596c0a83d177cd4c5 - [view on LGTM.com](https://lgtm.com/projects/g/ant-design/ant-design/rev/pr-20b59c4aecfda6ad22512e1a02c36358dd7a0a3a)

**new alerts:**

* 4 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by afc163** [comment]
This pull request **introduces 3 alerts** when merging 15ca3cb5dbe753444daca817595fc6e1bddfba0d into c4283ebe72fd79b17f9c5ff596c0a83d177cd4c5 - [view on LGTM.com](https://lgtm.com/projects/g/ant-design/ant-design/rev/pr-3932c76cdb9c882e4e5edf7254bf6036a73c16fd)

**new alerts:**

* 3 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by zombieJ** [comment]
https://deploy-preview-16026--ant-design.netlify.com/components/transfer-cn/#components-transfer-demo-customize

可以 review 起来了，看看 api 没问题我就开始搞用例了：https://github.com/ant-design/ant-design/pull/16026/files#diff-354b76c4434191794f8ee595d3cd16da

**Comment by afc163** [comment]
改成点击行选择如何，不然和默认的例子交互不一致。

File: components/transfer/demo/customize.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+---
+order: 5
+title:
+  zh-CN: 自定义列表
```
Comment (afc163) [inline components/transfer/demo/customize.md]: 表格穿梭框

File: components/transfer/demo/customize.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  en-US: Customize List
+---
+
+## zh-CN
+
+最基本的用法，展示了 `dataSource`、`targetKeys`、每行的渲染函数 `render` 以及回调函数 `onChange` `onSelectChange` `onScroll` 的用法。
```
Comment (afc163) [inline components/transfer/demo/customize.md]: 不是『最基本的用法』。

这句话代码太多了不容易读懂，要从需求角度重写下。

**Comment by zombieJ** [comment]
Update the test case. Please help to review.

**Comment by afc163** [comment]
例子里搞一个 disabled 的 item 就够了。

image

**Comment by afc163** [comment]
image

disabled 的时候不需要 hover 效果。

**Comment by afc163** [comment]
和之前的组件一样，你可以在 demo 里封装成一个 TransferTable 来调用。

File: components/transfer/demo/table-transfer.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  mockData.push({
+    key: i.toString(),
+    title: `content${i + 1}`,
+    description: `description of content${i + 1}`,
+    disabled: i % 3 < 1,
+    tag: mockTags[i % 3],
+  });
+}
+
+const oriTargetKeys = mockData
```
Comment (afc163) [inline components/transfer/demo/table-transfer.md]: originTargetKeys

File: components/transfer/list.tsx [hunk 4]
```diff
@@ -19,17 +19,25 @@ function isRenderResultPlainObject(result: any) {
...
   dataSource: TransferItem[];
-  filter: string;
   filterOption?: (filterText: any, item: any) => boolean;
   style?: React.CSSProperties;
   checkedKeys: string[];
   handleFilter: (e: any) => void;
   handleSelect: (selectedItem: any, checked: boolean) => void;
+  /** [Legacy] Only used when `body` prop used. */
   handleSelectAll: (dataSource: any[], checkAll: boolean) => void;
+  onItemSelect: (key: any, check: boolean) => void;
```
Comment (chenshuai2144) [inline components/transfer/list.tsx]: 这些 any 应该可以用泛型解决吧

File: components/transfer/index.zh-CN.md [hunk 5]
```diff
@@ -20,26 +20,42 @@ title: Transfer
...
+| targetKeys | 显示在右侧框数据的key集合 | string\[] | \[] | |
+| titles | 标题集合，顺序从左至右 | string\[] | \['', ''] | |
+| onChange | 选项在两栏之间转移时的回调函数 | (targetKeys, direction, moveKeys): void |  | |
+| onScroll | 选项列表滚动时的回调函数 | (direction, event): void |  | |
+| onSearch | 搜索框内容时改变时的回调函数 | (direction: 'left'\|'right', value: string): void | - | |
+| onSelectChange | 选中项发生改变时的回调函数 | (sourceSelectedKeys, targetSelectedKeys): void |  | |
+
+### Render Props
+
+3.17.0 新增。Transfer 支持接收 `children` 自定义渲染列表，并返回以下参数：
```
Comment (afc163) [inline components/transfer/index.zh-CN.md]: 写一点伪代码简单展示如何用？看整个 demo 还是有点复杂。

File: components/transfer/demo/table-transfer.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+          disabled={disabled}
+          showSearch={showSearch}
+          showSelectAll={false}
+          filterOption={(inputValue, item) => (
+            item.title.indexOf(inputValue) !== -1 ||
+            item.tag.indexOf(inputValue) !== -1
+          )}
+        >
+          {(listProps) => {
+            const { direction, filteredItems, onItemSelectAll, onItemSelect, selectedKeys: listSelectedKeys, disabled: listDisabled } = listProps;
```
Comment (afc163) [inline components/transfer/demo/table-transfer.md]: 直接放参数里解构出来好了。

File: components/transfer/demo/table-transfer.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                dataSource={filteredItems}
+                size="small"
+                onRow={({ key, disabled: itemDisabled }) => ({
+                  onClick: () => {
+                    if (itemDisabled || listDisabled) return;
+                    onItemSelect(key, !listSelectedKeys.includes(key));
+                  },
+                })}
+              />
+            );
```
Comment (afc163) [inline components/transfer/demo/table-transfer.md]: 如果要保持一边是原来的 list 可以怎么写？

Comment (zombieJ) [reply components/transfer/demo/table-transfer.md]: 那再搞一个一边是 Tree，一边是原来的？

File: components/transfer/list.tsx [hunk 8]
```diff
@@ -114,45 +118,49 @@ export default class TransferList extends React.Component<TransferListProps, any
   };
 
   handleClear = () => {
+    this.setState({ filterValue: '' });
     this.props.handleClear();
```
Comment (chenshuai2144) [inline components/transfer/list.tsx]: handleClear 是必填的吗？直接调用感觉有点慌

File: components/transfer/list.tsx [hunk 9]
```diff
@@ -19,17 +19,25 @@ function isRenderResultPlainObject(result: any) {
...
+  renderedEl: React.ReactNode;
+  item: TransferItem;
+}
+
 export interface TransferListProps {
   prefixCls: string;
   titleText: string;
   dataSource: TransferItem[];
-  filter: string;
   filterOption?: (filterText: any, item: any) => boolean;
```
Comment (chenshuai2144) [inline components/transfer/list.tsx]: filterOption?: (filterText:string, item: any) => boolean;

**Comment by chenshuai2144** [comment]
TransferList 的  searchPlaceholder，notFoundContent ，itemsUnit， itemUnit ，onScroll， direction
都是必填项，是历史遗留问题。一起改掉吧

File: components/transfer/demo/table-transfer.md (hunk 7 above)
Comment (afc163) [reply components/transfer/demo/table-transfer.md]: 嗯，这个需求很常见。

**Comment by zombieJ** [comment]
添加了一个树穿梭框的例子：https://deploy-preview-16026--ant-design.netlify.com/components/transfer-cn/#components-transfer-demo-tree-transfer

**Comment by zombieJ** [comment]
Update the TreeTransfer Demo, please check again.

**Review by afc163 — APPROVED** [review]

File: components/transfer/index.en-US.md [hunk 10]
```diff
@@ -18,28 +18,51 @@ One or more elements can be selected from either column, one click on the proper
...
+| targetKeys | A set of keys of elements that are listed on the right column. | string\[] | \[] | |
+| titles | A set of titles that are sorted from left to right. | string\[] | - | |
+| onChange | A callback function that is executed when the transfer between columns is complete. | (targetKeys, direction, moveKeys): void |  | |
+| onScroll | A callback function which is executed when scroll options list | (direction, event): void |  | |
+| onSearch | A callback function which is executed when search field are changed | (direction: 'left'\|'right', value: string): void | - | |
+| onSelectChange | A callback function which is executed when selected items are changed. | (sourceSelectedKeys, targetSelectedKeys): void |  | |
+
+### Render Props
+
+New in 3.17.0. Transfer accept `children` to customize render list, using follow props:
```
Comment (afc163) [inline components/transfer/index.en-US.md]: 错过站了

Comment (zombieJ) [reply components/transfer/index.en-US.md]: 😢 我更新一下。

**Comment by andyyxw** [comment]
这次更新后，`Transfer` 的 `ref` 暴露出的方法都用不了了。
比如 我要在其他组件触发 `Transfer` 的 `handleFilter` 来进行搜索。

**Comment by zombieJ** [comment]
@andyyxw, `handleFilter` 是内部方法，不应该通过 `ref` 去调用。如果有这类需求，可以来个 PR 把 search 变成受控属性。

**Pull request merged**